static/js/subs.js

```diff
@@ -523,17 +522,56 @@ function stream_matches_query(query, sub) {
 // query is now an object rather than a string.
 // Query { input: String, subscribed_only: Boolean }
 exports.filter_table = function (query) {
+    var stream_name_match_stream_ids = [];
+    var stream_description_match_stream_ids = [];
+    var others = [];
+    var stream_id_to_stream_name = {};
+    var widgets = {};
+
+    function sort_by_stream_name(a, b) {
+        var stream_a_name = stream_id_to_stream_name[a].toLocaleLowerCase();
+        var stream_b_name = stream_id_to_stream_name[b].toLocaleLowerCase();
+        return String.prototype.localeCompare.call(stream_a_name, stream_b_name);
+    }
+
     _.each($("#subscriptions_table .stream-row"), function (row) {
         var sub = stream_data.get_sub_by_id($(row).attr("data-stream-id"));
         sub.data_temp_view = $(row).attr("data-temp-view");
 
-        if (stream_matches_query(query, sub)) {
+        if (stream_matches_query(query, sub, 'name')) {
             $(row).removeClass("notdisplayed");
-        } else {
+
+            stream_id_to_stream_name[sub.stream_id] = sub.name;
+            stream_name_match_stream_ids.push(sub.stream_id);
+
+            widgets[sub.stream_id] = $(row).detach();
+        } else if (stream_matches_query(query, sub, 'description')) {
+            $(row).removeClass("notdisplayed");
+
+            stream_id_to_stream_name[sub.stream_id] = sub.name;
+            stream_description_match_stream_ids.push(sub.stream_id);
+
+            widgets[sub.stream_id] = $(row).detach();
+       } else {
             $(row).addClass("notdisplayed");
+            others.push($(row).detach());
         }
     });
 
+    stream_name_match_stream_ids.sort(sort_by_stream_name);
+    stream_description_match_stream_ids.sort(sort_by_stream_name);
```

Comment:
We should add a node test that verifies how the streams get sorted here, and we should have a test case that matches the keyword in both the stream name and the description to assure that we don't ever accidentally introduce duplicate rows in some refactoring.